File: widgets/mpris.js
const { Gtk, Gdk } = imports.gi;
const Mpris = await Service.import('mpris')

const media_controls = (player) => Widget.CenterBox({
    className: 'media-controls',
    hpack: 'center',
    hexpand: true,
    startWidget: Widget.Button({
        hpack: 'start',
        className: 'button',
        onClicked: () => player.previous(),
        child: Widget.Icon('media-skip-backward-symbolic')
    }),
    centerWidget: Widget.Button({
        hpack: 'center',
        className: 'button',
        onClicked: () => player.playPause(),
        child: Widget.Icon({
            setup: self => self.hook(Mpris, () => (player.playBackStatus === 'Playing') 
                ? self.icon = 'media-playback-pause-symbolic'
                : self.icon = 'media-playback-start-symbolic')
        })
    }),
    endWidget: Widget.Button({
        hpack: 'end',
        className: 'button',
        onClicked: () => player.next(),
        child: Widget.Icon('media-skip-forward-symbolic')
    })
})

const cover_with_controls = (player) => Widget.Box({
    className: "media",
    children: [
        Widget.Box({
            vertical: true,
            children: [
                media_controls(player),
                Widget.Slider({
                    vpack: 'end',
                    className: 'progress-bar',
                    drawValue: false,
                    hexpand: false,
                    onChange: ({ value }) => {player.position = (value * player.length)},
                    setup: self => self.poll(500, self => {
                        if (!player) return
                        self.value = player.position/player.length
                    })
                })
            ],
        }),
    ],
    setup: (self) => {
        self.hook(Mpris, () => {
            self.queue_draw()
            self.css = `background: linear-gradient(rgba(0, 0, 0, 0.6), rgba(0, 0, 0, 0.6)), url("${player.coverPath}"); min-height: 60px; min-width: 250px; background-position: 50% 50%; background-size: cover; border: 5px solid #161616`;
        })
    },
});


export const players = Widget.Stack({
    transition: "slide_up_down",
    transitionDuration: 125,
    children: {},
    setup: (self) => {
        self.add_events(Gdk.EventMask.SCROLL_MASK);
        self.add_events(Gdk.EventMask.SMOOTH_SCROLL_MASK);

        let currentDeltaY = 0;

        self.on("scroll-event", (_, event) => {
            const childNames = Object.keys(self.children);

            const length = Object.keys(self.children).length

            const prevChild = childNames[((n)=>n>=0?n:length-1)((childNames.indexOf(self.get_visible_child_name()) - 1) % length)];
            const nextChild = childNames[(childNames.indexOf(self.get_visible_child_name()) + 1) % length];

            const deltaY = event.get_scroll_deltas()[2];

            currentDeltaY += deltaY;

            if (currentDeltaY > 10 && prevChild) {
                self.set_visible_child_name(prevChild);
                currentDeltaY = 0;
            }
            if (currentDeltaY < -10 && nextChild) {
                self.set_visible_child_name(nextChild);
                currentDeltaY = 0;
            }
            console.log(self.children)
        });

        self.hook(Mpris, (_, name) => {
            if (!name) return;
            self.add_named(cover_with_controls(Mpris.getPlayer(name)), name);
        }, "player-added");
        self.hook(Mpris, (_, name) => {
            if (!name) return;

            self.get_child_by_name(name).destroy();
            delete children[name]
            console.log('destroyed')
        }, "player-closed");
    },
});

export const media = Widget.Revealer({
    revealChild: Mpris.bind("players").as((players) => players.length > 0),
    transition: "slide_up",
    transitionDuration: 125,
    child: players,
});

File: lua/widgets/bar/mpris.lua
local Mpris <const> = lgi.require("AstalMpris")
local mpris = Mpris.get_default()
local map = require('lua.lib').map

return Widget.Stack({
  transition_type = 'SLIDE_UP_DOWN',
  transition_duration = 125,
  children = {},
  setup = function(self)
    self:add_events(Gdk.EventMask.SCROLL_MASK);
    self:add_events(Gdk.EventMask.SMOOTH_SCROLL_MASK);

    local function add_player(player)
      self:add_named(
        Widget.Box({
          class_name = 'player',
          vertical = true,
          hexpand = false,
          setup = function(self) self['name'] = player:get_bus_name() end,
          css = bind(player, 'cover-art'):as(
            function(coverart)
              if coverart then 
                return 
                'background: linear-gradient(to right, rgba(0, 0, 0, 0), rgba(22, 22, 22, 0.9)), url("' .. coverart .. '");'
                .. 'background-position: 50% 50%; background-size: cover'
              else 
                return ''
              end
            end),
          children = {
            Widget.Box({
              vertical = true,
              hexpand = true,
              valign = 'CENTER',
              halign = 'END',
              children = {
                Widget.Button({
                  on_click = function() player:previous() end,
                  child = Widget.Icon({ icon = 'media-skip-backward-symbolic' })
                }),
                Widget.Button({
                  on_click = function() player:play_pause() end,
                  child = Widget.Icon({
                    icon = bind(player, 'playback-status'):as(function(status)
                      return (player:get_playback_status() == 'PLAYING' and 'media-playback-pause-symbolic' or 'media-playback-start-symbolic')
                    end)
                  })
                }),
                Widget.Button({
                  on_click = function() player:next() end,
                  child = Widget.Icon({ icon = 'media-skip-forward-symbolic' })
                })
              }
            }),
            Widget.Slider({
              hexpand = true,
              class_name = 'music-progress',
              valign = 'END',
              halign = 'START',
              -- for some reason, this is broken
              on_change_value = function(self) player:set_position(self:get_value() * player:get_length()) end,
              setup = function(self)
                Variable():poll(500, function()
                  if player == nil then return end
                  self:set_value(player:get_position() / player:get_length())
                end)
              end,
            })
          }
        }),
        player:get_bus_name()
      )
    end

    map(mpris:get_players(), add_player)

    self:hook(mpris, 'player-added', function(self, player)
      add_player(player)
    end)

    self:hook(mpris, 'player-closed', function(self, player)
      print('remove player', player:get_bus_name())
      self:get_named(player:get_bus_name()):destroy()
    end)

    local accumulated_delta_y = 0

    self:hook(self, 'scroll-event', function(_, event)
      local children = self:get_children()
      local current_child = self:get_visible_child()
      local _, delta_y = event:get_scroll_deltas()
      local index

      accumulated_delta_y = accumulated_delta_y + delta_y

      for i, child in ipairs(children) do 
        if child == current_child then
          index = i
        end
      end

      if math.abs(accumulated_delta_y) > 10 then
        self:set_visible_child(children[(index + math.floor(accumulated_delta_y / 10)) % #children + 1])
        accumulated_delta_y = 0
      end
    end)
  end
})
